### DISCUSSION TITLE
Peripheral Sound Problem. Please help.

### DISCUSSION THREAD

**Discussion by MAGGen-hub** [cc-tweaked/CC-Tweaked#886] [Q&A] [answered]
Dear developer of CC:Tweaked.

I get this problem while programming addon for your modification. My mod is very simple - advanced speaker peripherial with /stopsound comand support, 16 notes per tick and better volume control. I want to make it play paralel sounds:
Example: test.lua
peripherial.call("advanced_speaker_name","playSound","music_disc.pigstep")
peripherial.call("advanced_speaker_name","playSound","music_disc.11")
peripherial.call("advanced_speaker_name","playSound","music_disc.13")

My idea is: when you launch test.lua in shell all three sounds start playing imediatly (Need for custom instruments support from Note Block Studio that can be a little(or not little) biger than simple notes). But the main problem that SimpleChanel network can handle only one sound for one speaker at once. So what I need to do to make the count of paralel played sounds at least 8 (sorry for my weird looking english)? Possible Idea - create 8 different networks (NetworkHandler; NetworkHandler1; NetworkHandler2; ...) but I think it's not right... I know I can send all this sound to server.getPlayerList().broadcast( insert args here ) but how to move (their location in world) sounds launched from pocket computer to broadcast (I think method is exist but how to understand that THAT sound event launched from THIS poket comptuter and not from other sound source)? So please tell me what I can do or what I need to read to understand how your NetworkHandler works and understand how to improve it(any docs would be usefull).
"dan200.computercraft.shared.network.NetworkHandler". And I still can't understand

You created public static SimpleChannel network; for 1 sound event. How make it support 8 or 16 sound events? Is there em... any AdvancedChannel in forge mod loader? 

Thank you for your attantion and sorry for my bad english again. I hope you can help me. I can send you my code if you ask but all sound in it send to broadcast so I am unable to connect sound to pocket computer (case only your network sound position can be updated)...

P.S. (I got an Idea: maybe i need to create special id for each sound like UUID:id (getSource() + " " + Soundid) but I don't think that should work...)

**Comment by SquidDev** [comment, marked as answer]
I think it's worth noting here that none of CC's network code (or speaker code in general) is part of the public API. Nobody except CC itself is meant to be using it, hence the lack of documentation. You will need to reimplement a good chunk of this code in your own mod instead (ignore the whole `NetworkHandler` thing, it's just a wrapper over Forge's networking system).

> But the main problem that SimpleChanel network can handle only one sound for one speaker at once. 

This isn't really a limitation with SimpleChannel or anything to do with the network code, but rather with the `SoundManager` - this only allows one sound to be playing at a time. What you'd want is something similar to that, but which stores a map of `UUID`s to `List<MoveableSound>` (or possibly a set?).

Then, when playing a sound, you add to that list instead of stopping the previous sound and replacing it. You'd probably need some additional cleanup here to remove sounds once they're done.

**Reply by MAGGen-hub** [reply]
Thancks but I still cant understand (case can't test the server): SpeakerPlayClientMessage - plays message to all players in range or only for player that launched this message from HIS client? (sorry for stupid question)

**Reply by SquidDev** [reply]
Should be all players in range.

**Reply by MAGGen-hub** [reply]
From now I am searching for a way to make custom uuids like the main uuid part and the id's: speaker_uuid:sound_layer_id:sound_id. In noteblockstodio sounds separated by layers.. so I think it's the best identification system for speaker sounds...

**Reply by MAGGen-hub** [reply]
<Should be all players in range.>
Thanks... The word "client" allways scare me (what if it incomparable with server). The "flashback" from other games I modifed with their "client only methods"...

**Reply by MAGGen-hub** [reply]
I also crashed into volume problem. Have you seen any ways to edit REAL volume of sound (not the range in blocks)?

**Reply by MAGGen-hub** [reply]
Finaly... I understand how it works... And you right... I have no way to edit uuid after it creation... So I think this speaker need it's own events and it's own sound manager if I want it to work normaly... Em... Can I create my onw event message to insert into your network? If I can... then how can I do this? I mean how can I insert registry item into the setup method (or above) of your network without creating another network?

**Reply by SquidDev** [reply]
> I mean how can I insert registry item into the setup method (or above) of your network without creating another network?

You can't and you shouldn't. CC: Tweaked's network code is only meant to be used by CC: Tweaked - it is NOT meant to be used by another mod. I'm afraid you will need to reimplement a bunch of this functionality yourself (including the network code), but anything outside the `dan200.computercraft.api` package is internal to the mod and liable to change at any point.

>  The word "client" allways scare me (what if it incomparable with server).

When you create a `SimpleChannel`, you pass in a "current network version". You can then check the two versions match up - this way you don't need to worry about incompatibilities.

https://github.com/SquidDev-CC/CC-Tweaked/blob/01ddb2b4e436d5931159c25c15eafdfe33f64de1/src/main/java/dan200/computercraft/shared/network/NetworkHandler.java#L40-L41

---

> From now I am searching for a way to make custom uuids like the main uuid part and the id's: speaker_uuid:sound_layer_id:sound_id.

I feel I should clarify quite what the UUIDs are for here. Every speaker gets a unique identifier and when we play a sound we associate it with a specific uuid (and thus speaker). This means when a speaker is moved or broken, we can tell all the clients that the speaker/uuid has changed and the sound should be moved or stopped.

However, the fact that each speaker can only play one sound right now is entirely a client-side limitation. The client _could_ allow multiple sounds to be going on at once, I just didn't think it was the right choice for CC:T.

Obviously though you're going to need to write your own network code, so don't have to follow whatever CC:T has done! I'm sure there's plenty of better ways to achieve the same thing!

**Reply by MAGGen-hub** [reply]
Finaly! I finish the base of speaker (with my own network and sound manager). But I have another problem... How to show items like poket computers with speaker upgrade in my creative Tab(searching for solutions in Advanced Peripherial mod github case they already done it)?... Thank you for your help! Now I understand better how forge works...

**Reply by SquidDev** [reply]
You need to override `fillItemCategory` on one of your items to also add the turtle with your upgrade. For instance, this is what AdvancedPeripherals does:

https://github.com/Seniorendi/AdvancedPeripherals/blob/e7d1db2fcca61e75dd2edb08f6de27ddd6851a0d/src/main/java/de/srendi/advancedperipherals/common/items/base/BaseBlockItem.java#L57-L58
